File: apps/api/app/schemas/subscriptions.py
from __future__ import annotations

import datetime

from pydantic import BaseModel, ConfigDict, Field


class SubscriptionEvidenceItem(BaseModel):
    # The field is named `date`, so the type must be module-qualified or the
    # field name shadows it during pydantic's annotation evaluation.
    id: int | None = None
    date: datetime.date | None = None
    description: str | None = None
    amount: float | None = None


class SubscriptionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    merchant_key: str
    name: str
    category: str
    service_category: str
    amount: float
    monthly_amount: float
    annual_amount: float
    cycle: str
    cycle_days: int
    confidence: float
    status: str
    cancel_url: str | None = None
    replaceable: bool
    first_seen: datetime.date | None = None
    last_seen: datetime.date | None = None
    next_charge_date: datetime.date | None = None
    notes: str | None = None
    is_manual: bool
    cycle_is_manual: bool
    # validation_alias keeps from_attributes from ingesting the ORM's raw
    # `evidence` JSON string; the route layer assigns the parsed rows.
    evidence: list[SubscriptionEvidenceItem] = Field(default_factory=list, validation_alias="evidence_rows")


class SubscriptionSummary(BaseModel):
    active_count: int
    review_count: int
    action_count: int
    manage_link_count: int
    monthly_total: float
    annual_total: float
    potential_savings: float
    average_confidence: int
    transaction_count: int


class SubscriptionCategoryBreakdownRow(BaseModel):
    category: str
    amount: float
    percent: int


class SubscriptionOpportunity(BaseModel):
    subscription_id: int
    reason: str


class SubscriptionListResponse(BaseModel):
    subscriptions: list[SubscriptionResponse]
    summary: SubscriptionSummary
    category_breakdown: list[SubscriptionCategoryBreakdownRow]
    opportunities: list[SubscriptionOpportunity]
    upcoming_subscription_ids: list[int]
    statuses: dict[str, str]
    cycles: list[str]


class SubscriptionScanResponse(BaseModel):
    synced_count: int
    subscriptions: list[SubscriptionResponse]


class SubscriptionCreateRequest(BaseModel):
    name: str
    amount: float
    cycle: str = "Monthly"
    next_charge_date: str | None = None
    notes: str | None = None


class SubscriptionUpdateRequest(BaseModel):
    # Unified PATCH covering Flask's confirm/status/cycle/manage-url routes.
    status: str | None = None
    notes: str | None = None
    cycle: str | None = None
    cancel_url: str | None = None


class SubscriptionLinkHelpRequest(BaseModel):
    pass


class SubscriptionLinkCandidateResponse(BaseModel):
    title: str
    url: str
    reason: str
    confidence: str


class SubscriptionLinkHelpResponse(BaseModel):
    source: str
    provider: str
    model: str
    status: str
    candidates: list[SubscriptionLinkCandidateResponse] = Field(default_factory=list)
    message: str


class SubscriptionImportRequest(BaseModel):
    csv_text: str | None = None
    csv_base64: str | None = None


class SubscriptionImportResponse(BaseModel):
    imported: int
    synced_count: int

File: apps/web/lib/subscriptions.ts
import type { components } from "@clearpath/api-client";
import { subscriptionsViewSchema } from "@clearpath/validation";

type ApiMe = components["schemas"]["MeResponse"];
type ApiSubscriptions = components["schemas"]["SubscriptionListResponse"];

function mapSession(data: ApiMe) {
  return {
    ownerUserId: data.id,
    householdName: data.household_name ?? null,
    selectedPlan: data.selected_plan,
    billingStatus: data.billing_status,
    planDisplayName: data.plan_display_name,
    primaryAccountHolder: data.primary_account_holder,
    subject: {
      id: data.session_subject.id,
      subjectType: data.session_subject.subject_type,
      email: data.session_subject.email,
      displayName: data.session_subject.display_name,
      firstName: data.session_subject.first_name,
      avatarInitial: data.session_subject.avatar_initial,
      householdRole: data.session_subject.household_role ?? null,
    },
    featureAccess: (data.feature_access ?? []).map((row) => ({
      feature: row.feature,
      enabled: row.enabled,
      hidden: row.hidden,
      requiredPlan: row.required_plan,
    })),
  };
}

export function mapSubscriptions(data: ApiSubscriptions, me: ApiMe) {
  return subscriptionsViewSchema.safeParse({
    session: mapSession(me),
    subscriptions: data.subscriptions.map((row) => ({
      id: row.id,
      merchantKey: row.merchant_key,
      name: row.name,
      category: row.category,
      serviceCategory: row.service_category,
      amount: row.amount,
      monthlyAmount: row.monthly_amount,
      annualAmount: row.annual_amount,
      cycle: row.cycle,
      cycleDays: row.cycle_days,
      confidence: row.confidence,
      status: row.status,
      cancelUrl: row.cancel_url ?? null,
      replaceable: row.replaceable,
      firstSeen: row.first_seen ?? null,
      lastSeen: row.last_seen ?? null,
      nextChargeDate: row.next_charge_date ?? null,
      notes: row.notes ?? null,
      isManual: row.is_manual,
      cycleIsManual: row.cycle_is_manual,
      evidence: (row.evidence ?? []).map((item) => ({
        id: item.id ?? null,
        date: item.date ?? null,
        description: item.description ?? null,
        amount: item.amount ?? null,
      })),
    })),
    summary: {
      activeCount: data.summary.active_count,
      reviewCount: data.summary.review_count,
      actionCount: data.summary.action_count,
      manageLinkCount: data.summary.manage_link_count,
      monthlyTotal: data.summary.monthly_total,
      annualTotal: data.summary.annual_total,
      potentialSavings: data.summary.potential_savings,
      averageConfidence: data.summary.average_confidence,
      transactionCount: data.summary.transaction_count,
    },
    categoryBreakdown: data.category_breakdown.map((row) => ({ category: row.category, amount: row.amount, percent: row.percent })),
    opportunities: data.opportunities.map((row) => ({ subscriptionId: row.subscription_id, reason: row.reason })),
    upcomingSubscriptionIds: data.upcoming_subscription_ids,
    statuses: data.statuses,
    cycles: data.cycles,
  });
}
